bin/transform-readme.php

```diff
@@ -0,0 +1,178 @@
+
+				if ( 0 !== strpos( $license_uri, 'http' ) ) {
+					fwrite( STDERR, "Unable to extract License URI from: $meta.\n" );
+					exit( __LINE__ );
+				}
+
+				$metadata['License URI'] = $license_uri;
+			}
+		}
+
+		$expected_metadata = [
+			'Contributors',
+			'Tags',
+			'Requires at least',
+			'Tested up to',
+			'Stable tag',
+			'License',
+			'License URI',
+			'Requires PHP',
+		];
+		foreach ( $expected_metadata as $key ) {
+			if ( empty( $metadata[ $key ] ) ) {
+				fwrite( STDERR, "Failed to parse metadata. Missign: $key\n" );
+				exit( __LINE__ );
+			}
+		}
+
+		$replaced = "$header\n";
+		foreach ( $metadata as $key => $value ) {
+			$replaced .= "$key: $value\n";
+		}
+		$replaced .= "\n$description\n\n";
+
+		return $replaced;
+	},
+	$readme_txt
+);
+
+// Replace image-linked YouTube videos with bare URLs.
+$readme_txt = preg_replace(
+	'#\[!\[.+?]\(.+?\)]\((https://www\.youtube\.com/.+?)\)#',
+	'$1',
+	$readme_txt
+);
+
+// Fix up the screenshots.
+$screenshots_captioned = 0;
+$readme_txt            = preg_replace_callback(
+	'/(?<=## Screenshots\n\n)(.+?)(?=## Changelog)/s',
+	static function ( $matches ) use ( &$screenshots_captioned ) {
+		if ( ! preg_match_all( '/### (.+)/', $matches[0], $screenshot_matches ) ) {
+			fwrite( STDERR, "Unable to parse screenshot headings.\n" );
+			exit( __LINE__ );
+		}
+
+		$screenshot_txt = '';
+		foreach ( $screenshot_matches[1] as $i => $screenshot_caption ) {
+			$screenshot_txt .= sprintf( "%d. %s\n", $i + 1, $screenshot_caption );
+			$screenshots_captioned++;
+		}
+		$screenshot_txt .= "\n";
+
+		return $screenshot_txt;
+	},
+	$readme_txt,
+	1,
+	$replace_count
+);
+if ( 0 === $replace_count ) {
+	fwrite( STDERR, "Unable to transform screenshots.\n" );
+	exit( __LINE__ );
+}
+
+$screenshot_files = glob( __DIR__ . '/../.wordpress-org/screenshot-*' );
+if ( count( $screenshot_files ) !== $screenshots_captioned ) {
+	fwrite( STDERR, "Number of screenshot files does not match number of screenshot captions.\n" );
+	exit( __LINE__ );
+}
+foreach ( $screenshot_files as $i => $screenshot_file ) {
+	if ( 0 !== strpos( basename( $screenshot_file ), sprintf( 'screenshot-%d.', $i + 1 ) ) ) {
+		fwrite( STDERR, "Screenshot filename is not sequential: $screenshot_file.\n" );
+		exit( __LINE__ );
+	}
+}
+
+// Convert markdown headings into WP readme headings for good measure.
+$readme_txt = preg_replace_callback(
+	'/^(#+)\s(.+)/m',
+	static function ( $matches ) {
+		$md_heading_level = strlen( $matches[1] );
+		$heading_text     = $matches[2];
+
+		// #: ===
+		// ##: ==
+		// ###: =
+		$txt_heading_level = 4 - $md_heading_level;
+		if ( $txt_heading_level <= 0 ) {
+			fwrite( STDERR, "Heading too small to transform: {$matches[0]}.\n" );
+			exit( __LINE__ );
+		}
+
+		return sprintf(
+			'%1$s %2$s %1$s',
+			str_repeat( '=', $txt_heading_level ),
+			$heading_text
+		);
+	},
+	$readme_txt,
+	-1,
+	$replace_count
+);
+if ( 0 === $replace_count ) {
```

Comment:
This condition will never be true as `$replace_count` will always have a value > 0 after the screenshot transformations above, and `preg_replace_callback()` will increase that previous value by the number of replacements done here. So, the condition should either be `$replace_count === $replace_count`, or use an unset variable.

Reply:
This doesn't seem to be the case: https://3v4l.org/0CGoT

```php
<?php
$str = 'abcba';

$str = preg_replace( '/b/', 'a', $str, -1, $replace_count );
var_dump( $replace_count );

$str = preg_replace( '/c/', 'a', $str, -1, $replace_count );
var_dump( $replace_count );

var_dump( $str );
```

The result is:

```
int(2)
int(1)
string(5) "aaaaa"
```

In other words, `$replace_count` is reset with each `preg_replace_callback()`. It doesn't increment on top of an existing variable. If it did, then the `int(3)` would be here instead of `int(1)`.

Reply:
Yea I've been terribly mistaken here. I was operating against changes I made locally :facepalm:. This is all good.